File: csharp_solutions/easy/InvertBinaryTree.cs
namespace InvertBinaryTree
{
    public class TreeNode
    {
        public int val;
        public TreeNode? left;
        public TreeNode? right;
        public TreeNode(int val = 0, TreeNode? left = null, TreeNode? right = null)
        {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public class Solution
    {
        public TreeNode? InvertTree(TreeNode? root)
        {
            if (root is null)
            {
                return root;
            }
            TreeNode? tempRoot = root.right;
            root.right = root.left;
            root.left = tempRoot;

            InvertTree(root.right);
            InvertTree(root.left);

            return root;
        }
    }
}

File: 02_session/easy/invert_binary_tree.rb
# frozen_string_literal: true

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {TreeNode}
def invert_tree(root)
  return root if root.nil?

  root.left, root.right = root.right, root.left

  invert_tree(root.left)
  invert_tree(root.right)
  root
end

# 5min

# NOTES
# algo/structure: binary tree, DFS
# time: O(n)
# space: O(n)

# simple DFS problem
